matchers/parse/parse.go

```diff
@@ -0,0 +1,315 @@
+// Copyright 2023 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package parse
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/alertmanager/pkg/labels"
+)
+
+var (
+	ErrEOF          = errors.New("end of input")
+	ErrNoOpenBrace  = errors.New("expected opening brace")
+	ErrNoCloseBrace = errors.New("expected close brace")
+	ErrNoLabelName  = errors.New("expected label name")
+	ErrNoLabelValue = errors.New("expected label value")
+	ErrNoOperator   = errors.New("expected an operator such as '=', '!=', '=~' or '!~'")
+	ErrExpectedEOF  = errors.New("expected end of input")
+)
+
+// Parser reads the sequence of tokens from the lexer and returns either a
+// series of matchers or an error. An error can occur if the lexer attempts
+// to scan text that does not match the expected grammar, or if the tokens
+// returned from the lexer cannot be parsed into a complete series of matchers.
+// For example, the input is missing an opening bracket, has missing label
+// names or label values, a trailing comma, or missing closing bracket.
+type Parser struct {
+	done         bool
+	err          error
+	hasOpenParen bool
+	input        string
+	lexer        Lexer
+	matchers     labels.Matchers
+}
+
+func NewParser(input string) Parser {
+	return Parser{
+		input: input,
+		lexer: NewLexer(input),
+	}
+}
+
+// Error returns the error that caused parsing to fail.
+func (p *Parser) Error() error {
+	return p.err
+}
+
+// Parse returns a series of matchers or an error. It can be called more than
+// once, however successive calls return the matchers and err from the first
+// call.
+func (p *Parser) Parse() (labels.Matchers, error) {
+	if !p.done {
+		p.done = true
+		p.matchers, p.err = p.parse()
+	}
+	return p.matchers, p.err
+}
+
+// accept returns true if the next token is one of the expected kinds, or
+// an error if the next token that would be returned from the lexer does not
+// match the expected grammar, or if the lexer has reached the end of the input
+// and TokenNone is not one of the accepted kinds. It is possible to use either
+// Scan() or Peek() as fn depending on whether accept should consume or peek
+// the next token.
+func (p *Parser) accept(fn func() (Token, error), kind ...TokenKind) (bool, error) {
```

Comment:
I understand the need to try and keep some level of parity between `accept` and `expect` but I don't see the correlation between them.

My understanding of the current usage of `accept` is `Peek` at the next token and tell me if it's the kind I expect. As such, it feels clear to me to just inline this as it's only relevant on the `parseOpenParen` function and there's no need for an extraction.


I propose the following:

```golang
// In parse.go
func (p *Parser) PeekNext() (Token, error) {
	t, err := p.lexer.Peek()
	if err != nil {
		return Token{}, err
	}

	if t.Kind == TokenNone {
		return Token{}, fmt.Errorf("0:%d: %w", len(p.input), ErrEOF)
	}

	return t, nil
}
// in token.go
// IsAny verifies that the token is of any specified TokenKinds.
func (t Token) IsAny(kinds ...TokenKind) bool {
	for _, k := range kinds {
		if t.Kind == k {
			return true
		}
	}

	return false
}
```

And finally, change `parseOpenParen`

```golang
func (p *Parser) parseOpenParen(l *Lexer) (parseFn, error) {
	// Can start with an optional open brace.
	currentToken, err := p.peekNext()
	if err != nil {
		if errors.Is(err, ErrEOF) {
			return p.parseEOF, nil
		}
		return nil, err
	}

	p.hasOpenParen = currentToken.IsAny(TokenOpenBrace)
	// If the token was an open brace it must be scanned so the token
	// following it can be peeked.
	if p.hasOpenParen {
		if _, err = l.Scan(); err != nil {
			panic("Unexpected error scanning open brace")
		}

		// If the next token is a close brace there are no matchers in the input,
		// and we can just parse the close brace.
		currentToken, err = p.peekNext()
		if err != nil {
			return nil, fmt.Errorf("%s: %w", err, ErrNoCloseBrace)
		}

		if currentToken.IsAny(TokenCloseBrace) {
			return p.parseCloseParen, nil
		}
	}

	if currentToken.IsAny(TokenCloseBrace) {
		return p.parseCloseParen, nil
	}

	return p.parseLabelMatcher, nil
}
```

Reply:
Updated. I'm thinking if I should just update the lexer to return an `ErrorEOF` error at the same time as `TokenNone` if there is no more input. I think doing so would remove the need for `peekNext` and instead we can just call `l.Peek()`.

Reply:
Discussed offline, agreed that we'll move the `ErrorEOF` errors into the lexer to simplify what's going on the parser. I'll wait on that before I continue with the review.